### PR TITLE
Editor: Remove unused Class Name field from Create Script dialog

### LINKED ISSUE #208: Allow setting of GDScript `class_name` property through the Create Script menu
**Describe the project you are working on:**

I am working on a program that allows the user to configure a thermometer-style progress bar to display.

**Describe the problem or limitation you are having in your project:**

I noticed that when creating GDScript files, the "Class Name" option in the menu is locked. The docs say this is because "the file itself represents the class". However, recently the `class_name` property was added, and I think this menu was overlooked. My "problem" is not as much a problem as it is a notice of inconsistency and overlook within the editor.

**Describe how this feature / enhancement will help you overcome this problem or limitation:**

 My proposal is to unlock this field and allow setting `class_name` from the menu, similarly to how `extends` can be set using the "Inherits" field. This would fix the inconsistency and overlook mentioned above.

**Show a mock up screenshots/video or a flow diagram explaining how your proposal will work:**

h

**Describe implementation detail for your proposal (in code), if possible:** N/A

**If this enhancement will not be used often, can it be worked around with a few lines of script?:**

It can be worked around with one line; however, it should in theory be used just as often as the `class_name` keyword itself if implemented. It is also an easy fix to an inconsistency.

**Is there a reason why this should be core and not an add-on in the asset library?:**

This is not a "new feature" like those found in the asset library; it is just an expansion of an existing feature that makes logical sense.

### LINKED ISSUE #75071: When creating a new script, Class Name is always "N/A" preventing auto-generation of class_name
### Godot version

v4.0.stable.official [92bee43ad]

### System information

Linux Ubuntu 22.04 with Unity desktop

### Issue description

I always have to write my class_name myself, so I thought a setting to auto-generate class name from the script file name, or at least to manually input it at script file creation, would be nice.

I found the field I wanted in the Create Script popup as "Class Name", except the content is filled with N/A no matter what. It doesn't auto-fill from file name, and I cannot edit the field manually either.

I'm using Godot without C#. Is it only active in C#? Either way, it is still useful in GDScript to auto-fill class_name at the top.

image

First I'd like to see the expected behavior... If it's only meant for manual fill, then once it's fixed, I will submit a suggestion to go further and auto-fill from file name.

_And forgive me if it was already reported. Turns out that searching "N/A" is super hard when so many issues set "Minimal reproduction project" as "N/A"... including this one!_

### Steps to reproduce

1. Create a new script
2. In the Create Script popup, note how Class Name is "N/A".
3. Try to edit the field, to no avail.
4. Edit the path, but Class Name is still not updated.

### Minimal reproduction project

N/A

### PR DESCRIPTION
**Pull request by dalexeev**
It looks like the field hasn't done anything for a very long time, either with GDScript or C# ([link](https://github.com/godotengine/godot/pull/12365/files#diff-42f5e0775f4f420d4c05168dc6db0c61d58ea2fda2c7481a5ddc9df4ab4f3955L296-R301)).

* Closes #75071.
* Closes godotengine/godot-proposals#208.

### CODE REVIEW

**Comment by adamscott** [comment]
What happens when a user don't supply a class name?

**Comment by AThousandShips** [comment]
The template line is removed, at least that's how it seems in the source

**Review by adamscott — APPROVED** [review]
As this would need thorough testing (as this is something that is used frequently), I suggest to merge this PR right after the 4.1 release. We will be able to get some feedback from the users if this PR is alright.

So, you get my pass. Thanks @dalexeev!

**Comment by dalexeev** [comment]
> What happens when a user don't supply a class name?

Good point. The field is not filled in automatically. If the user leaves it empty, then a script without `class_name` will be generated. I think we should add an "Optional" placeholder to make this more obvious.

> I suggest to merge this PR right after the 4.1 release

I'm not rushing anyone with this, sorry if I'm distracting. I just did not find today how I can help with 4.1.

**Comment by adamscott** [comment]
> I'm not rushing anyone with this, sorry if I'm distracting. I just did not find today how I can help with 4.1.

Don't worry. @akien-mga said that he was happy to have PRs ready to merge right after the release of 4.1.

**Comment by akien-mga** [comment]
Yeah it's good to think about the UX a bit and the implications here. If the UI encourages users to name all their classes, they might end up with hundreds of class names where it's not needed.

**Comment by adamscott** [comment]
> If the UI encourages users to name all their classes, they might end up with hundreds of class names where it's not needed.

The UI could say something like `- A global class named "MyClass" will be created`

File: editor/script_create_dialog.cpp [hunk 1]
```diff
@@ -304,16 +281,8 @@ String ScriptCreateDialog::_validate_path(const String &p_path, bool p_file_must
...
-	if (has_named_classes) {
-		return class_name->get_text();
-	} else {
-		return ProjectSettings::get_singleton()->localize_path(file_path->get_text()).get_file().get_basename();
-	}
-}
-
 void ScriptCreateDialog::_class_name_changed(const String &p_name) {
-	is_class_name_valid = _validate_class(class_name->get_text());
+	is_class_name_valid = class_name->get_text().is_empty() || class_name->get_text().is_valid_identifier();
```
Comment (adamscott) [inline editor/script_create_dialog.cpp:285]: Maybe we should check if the class has a name already used by a native or a global class. Then, we could output a custom error message about that.

**Comment by aaronfranke** [comment]
Is this really the right move though? I think it would be better to remove this field. It's easy for the user to write `class_name` manually. Having this in the UI adds clutter and encourages users to name all of their classes, which is often not desired since it will fill the Create New Node dialog with tons of useless entries.

**Comment by dalexeev** [comment]
There is an issue and a proposal about it, and there is also a similar proposal about the node name field in the node creation dialog, so I thought that users want this feature. But given the possible misuse of this field, I agree with you that it's better to remove it.

**Comment by dalexeev** [comment]
I have updated the PR. But it is possible that in the process we can return to the old version or come to something else.

**Comment by akien-mga** [comment]
Be careful removing the feature, AFAIK it's required for C# scripts. Any C# class needs to have a unique name (but it's not the same as the global script classes mechanism).

**Comment by dalexeev** [comment]
I haven't tested this, but for C# nothing should change, the class name is generated from the file name as before.

**Review by aaronfranke — COMMENTED** [review]
I support this change, but I won't give it an approval because there are some other possibly unrelated changes in this PR like changes to the script templates, and I don't know if those are desired to have in the same PR or if they should be split off (or maybe, the removal of the unused field should be the one split off).

For C#, it is generally expected that the class name should match the file name, so I don't think we need a separate field for specifying the class name.

**Review by akien-mga — APPROVED** [review]

**Pull request merged**